### ISSUE TITLE
Did not reach requested beam in determinize-lattice: size exceeds maximum 50000000 bytes; (repo,arcs,elems)

### ISSUE THREAD

**Issue by whaozl** [nvidia-riva/riva-asrlib-decoder#42]
Could you please help me take a look at these two issues? What are the main reasons causing and leading to these two problems, and what are the possible solutions?

```shell

CheckMemoryUsage():lat/determinize-lattice-pruned.cc:320) Did not reach requested beam in determinize-lattice: size exceeds maximum 50000000 bytes; (repo,arcs,elems) = (21918336,5826912,22257216), after rebuilding, repo size was 13777184, effective beam was 2.41899 vs. requested beam 8
WARNING ([5.5]:DeterminizeLatticePruned():lat/determinize-lattice-pruned.cc:1284) Effective beam 2.41899 was less than beam 8 * cutoff 0.5, pruning raw lattice with new beam 4.39908 and retrying.


WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2083) Resized host vector containing 'acoustic costs'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984
WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2100) Resized host vector containing 'infotoken'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984

```

`Did not reach requested beam in determinize-lattice: size exceeds maximum`  It will cause the WFST decode to be very slow and yield no prediction results.

**Comment by galv** [comment]
Regarding

```
WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2083) Resized host vector containing 'acoustic costs'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984
WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2100) Resized host vector containing 'infotoken'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984
```

This happens before determinization, so we can look at it first. Basically, you need to increase `ntokens_pre_allocated` to be a higher value than 1000000.

To give a sense of the memory usage implied by this, these are the vectors allocated with size ntokens_pre_allocated:

```
  std::vector<std::vector<InfoToken>> h_all_tokens_info_;
  std::vector<std::vector<CostType>> h_all_tokens_acoustic_cost_;
  std::vector<std::vector<InfoToken>> h_all_tokens_extra_prev_tokens_;
  std::vector<std::vector<float2>>
      h_all_tokens_extra_prev_tokens_extra_and_acoustic_cost_;
```

Basically, the first vector is indexed by channel id, and the second vector is indexed by ntokens_pre_allocated (but can grow in size if necessary). So it will imply host memory usage of:

`ntokens_pre_allocated * num_channels * (sizeof(InfoToken) + sizeof(CostType) + sizeof(InfoToken) + sizeof(float2))`, which is `ntokens_pre_allocated * num_channels * (8 + 4 + 8 + 8)` bytes.

Therefore, setting ntokens_pre_allocated too high to avoid this issue can be problematic for memory usage, even if it is unused, since unfortunately std::vector's constructor will allocate physical pages during initialization when it runs the constructor for each element in the vector.

With num_channels = 400, ntokens_pre_allocated=1000000, you would get 1000000 * 400 * (8 + 4 + 8 + 8) / 1024 / 1024/ 1024 = 10.4 GiB of host memory used. Doubling to ntokens_pre_allocated=2000000 is probably safe, but still a little bit annoying.

I mention these numbers because they are the default configs here:

https://github.com/nvidia-riva/riva-asrlib-decoder/blob/c94c84efb3efb526ce87fa9728a3dd3e621bb484/src/riva/asrlib/decoder/test_graph_construction.py#L784-L787

Note that, if you are doing offline decoding only (i.e., your are feeding the entire input audio in at once), you can safely do:

`config.online_opts.num_channels = config.online_opts.max_batch_size`

`config.online_opts.num_channels = 2 * config.online_opts.max_batch_size` is helpful only for online decoding, where we must switch between different audio streams while waiting on new audio to come in from existing streams to fully utilize the GPU.

**Comment by galv** [comment]
Regarding your other error:

```
CheckMemoryUsage():lat/determinize-lattice-pruned.cc:320) Did not reach requested beam in determinize-lattice: size exceeds maximum 50000000 bytes; (repo,arcs,elems) = (21918336,5826912,22257216), after rebuilding, repo size was 13777184, effective beam was 2.41899 vs. requested beam 8
WARNING ([5.5]:DeterminizeLatticePruned():lat/determinize-lattice-pruned.cc:1284) Effective beam 2.41899 was less than beam 8 * cutoff 0.5, pruning raw lattice with new beam 4.39908 and retrying.
```

Using a subtly incorrect FST for training can sometimes cause this, but I doubt that is your problem since WFST creation has become pretty stable at this point.

You can try increasing the max memory, which is defined here: https://github.com/galv/kaldi/blob/bbee966746fff1a74a3924a2c7ce4a611553a47c/src/lat/determinize-lattice-pruned.h#L145-L162

And modifiable from riva-asrlib-decoder here: https://github.com/nvidia-riva/riva-asrlib-decoder/blob/c94c84efb3efb526ce87fa9728a3dd3e621bb484/src/riva/asrlib/decoder/python_decoder.cc#L94

However, that has the problem of increasing runtime and memory usage. Here is more general advise:

Generally speaking, the memory usage will go up linearly with larger audio files. See here: https://groups.google.com/g/kaldi-help/c/yNdfPGDPifI/m/yi9Ejz4CDQAJ

If you are trying to decoder audio files larger than 30 seconds or 1 minute in length, I would encourage you to split the audio up into chunks. You can either used VAD or you can split arbitrarily. Then you would need to "restitch" the data. I personally have not made an easy way to do this in the riva-asrlib-decoder library, though. You can see how kaldi did it if you follow this code. though it is a little confusing: https://github.com/kaldi-asr/kaldi/pull/4553/files#diff-c37f1f54a80cd1b55a711157a8b9d56d975294f62b1bef6c9df87ff55480625eR67-R70

If you don't want to bother with splitting up your data, simply avoiding determinization is a good way to reduce memory pressure problems. You can use the "online decoder" in a fully offline mode, which will simply find the 1-best path by a simple backtracking, bypassing. See my code here: https://github.com/nvidia-riva/riva-asrlib-decoder/blob/c94c84efb3efb526ce87fa9728a3dd3e621bb484/src/riva/asrlib/decoder/test_graph_construction.py#L1420-L1432 The con is that this gives you only the one-best transcript, not a lattice containing the n-best transcripts, but that is okay for a lot of use cases. Note that this simply gets rid of the determinization error you received; it does not get rid of these warnings:

```
WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2083) Resized host vector containing 'acoustic costs'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984
WARNING ([5.5]:ComputeH2HCopies():cudadecoder/cuda-decoder.cc:2100) Resized host vector containing 'infotoken'. This can cause latency spikes in production. Please use smaller input audio sequences or increase ntokens_pre_allocatedOld capacity: 1000000, New capacity:1995984
```

I hope this helps. Let me know if you have questions.

**Comment by whaozl** [comment]
Thank you for providing such a detailed answer to my question.

and last, I find the problem is:

the model export to onnx, the onnx include the softmax layer(actually the onnx model cann't the softmax layer).

becase the input rivalib decode must log10

**Comment by whaozl** [comment]
https://github.com/nvidia-riva/riva-asrlib-decoder/blob/main/src/riva/asrlib/decoder/batched-mapped-decoder-cuda.cc#L21

```shell
const float kSleepForNewTask = 100e-6;
```

I would like to ask if this value is reasonable, and whether it might lead to excessive CPU utilization. Should the normal code sleep at least for 2 milliseconds?

If set to 2 milliseconds, would it cause a delay in fetching data in time, resulting in batch inference issues? Process as individual item to infer in the end?

**Comment by galv** [comment]
> becase the input rivalib decode must log10

You can try removing the softmax layer with onnx graph surgeon: https://github.com/NVIDIA/TensorRT/tree/master/tools/onnx-graphsurgeon/examples/06_removing_nodes

In general, I don't recommend export to onnx these days. I consider pytorch on its own to be much more flexible and easier to debug performance problems in. However, I am also a core contributor to pytorch, which makes it easier for me to fix things inside of it. The nemo results shown here use pure pytorch, with no onnx export: https://developer.nvidia.com/blog/accelerating-leaderboard-topping-asr-models-10x-with-nvidia-nemo/

> I would like to ask if this value is reasonable, and whether it might lead to excessive CPU utilization. Should the normal code sleep at least for 2 milliseconds?

Ideally it should be using condition variables, rather than sleeping for a specific amount of time. I recall finding a problem with address sanitizer when I tried to replace that sleep with a condition variable, though.

The original codebase in kaldi used a lot of sleep durations that were hand tuned for Turing generation GPUs. Those durations were very incorrect.

Anyway, sleeping for 100 microseconds is a fine compromise in my opinion. The thread is waking up only to check only two booleans. It is much worse to have a long sleep period like 2 ms since that increases GPU idle time by up to almost 4ms in the worst case.

**Comment by whaozl** [comment]
The nemo results shown here use pure pytorch, with no onnx export: [https://developer.nvidia.com/blog/accelerating-leaderboard-topping-asr-models-10x-with-nvidia-nemo/](https://developer.nvidia.com/blog/accelerating-leaderboard-topping-asr-models-10x-with-nvidia-nemo/)

I will try it to my asr model. Thank you for your help.
